lib/src/core/atom_derivation.dart

```diff
@@ -22,76 +16,81 @@ class Atom {
     }
   }
 
+  String name;
+
+  bool isPendingUnobservation = false;
+
+  DerivationState lowestObserverState = DerivationState.notTracking;
+
+  bool isBeingObserved = false;
+
   Set<Derivation> observers = Set();
 
-  var _observationListeners = Map<String, Set<Function>>();
+  final Map<_ListenerKind, Set<Function()>> _observationListeners = {};
 
-  reportObserved() {
+  void reportObserved() {
     ctx.reportObserved(this);
   }
 
-  reportChanged() {
-    ctx.startBatch();
-    ctx.propagateChanged(this);
-    ctx.endBatch();
+  void reportChanged() {
+    ctx
+      ..startBatch()
+      ..propagateChanged(this)
+      ..endBatch();
   }
 
-  addObserver(Derivation d) {
+  void addObserver(Derivation d) {
     observers.add(d);
 
     if (lowestObserverState.index > d.dependenciesState.index) {
       lowestObserverState = d.dependenciesState;
     }
   }
 
-  removeObserver(Derivation d) {
+  void removeObserver(Derivation d) {
     observers.removeWhere((ob) => ob == d);
     if (observers.isEmpty) {
       ctx.enqueueForUnobservation(this);
     }
   }
 
-  notifyOnBecomeObserved() {
-    if (_observationListeners[ON_BECOME_OBSERVED] != null) {
-      _observationListeners[ON_BECOME_OBSERVED]
-          .forEach((listener) => listener());
-    }
+  void notifyOnBecomeObserved() {
+    final listeners = _observationListeners[_ListenerKind.onBecomeObserved];
+    listeners?.forEach(_notifyListener);
   }
 
-  notifyOnBecomeUnobserved() {
-    if (_observationListeners[ON_BECOME_UNOBSERVED] != null) {
-      _observationListeners[ON_BECOME_UNOBSERVED]
-          .forEach((listener) => listener());
-    }
-  }
+  static void _notifyListener(Function() listener) => listener();
 
-  void Function() onBecomeObserved(Function fn) {
-    return _addListener(ON_BECOME_OBSERVED, fn);
+  void notifyOnBecomeUnobserved() {
+    final listeners = _observationListeners[_ListenerKind.onBecomeUnobserved];
+    listeners?.forEach(_notifyListener);
   }
 
-  void Function() onBecomeUnobserved(Function fn) {
-    return _addListener(ON_BECOME_UNOBSERVED, fn);
-  }
+  void Function() onBecomeObserved(Function fn) =>
+      _addListener(_ListenerKind.onBecomeObserved, fn);
+
+  void Function() onBecomeUnobserved(Function fn) =>
+      _addListener(_ListenerKind.onBecomeUnobserved, fn);
 
-  void Function() _addListener(String key, Function fn) {
+  void Function() _addListener(_ListenerKind kind, Function fn) {
     if (fn == null) {
-      throw MobXException('${key} handler cannot be null');
+      throw MobXException('$kind handler cannot be null');
```

Comment:
I think using `$kind` might just show the number `0` or `1` and may not be that helpful. This was the reason for choosing explicit strings. Is there a quick enum -> string conversion possible for a better exception message?

Reply:
This would be formatted as  "_ListenerKind.onBecomeObserved handler cannot be null". Not ideal, but enum toString is not just the number in Dart.

Reply:
That is good enough for now. Not sure if we can override `toString()` for an enum

Reply:
Unfortunately no, enums can't have methods or anything. They're quite crippled compared to Java enum classes etc.

Reply:
Ya that would been a great addition to the language. Hoping it would come up in future versions of dart